concrete/src/Permission/IpAccessControlService.php

```diff
@@ -0,0 +1,480 @@
+use Concrete\Core\Logging\LoggerAwareTrait;
+use DateTime;
+use Doctrine\Common\Collections\Criteria;
+use Doctrine\ORM\EntityManagerInterface;
+use IPLib\Address\AddressInterface;
+use IPLib\Factory as IPFactory;
+use IPLib\Range\RangeInterface;
+
+class IpAccessControlService implements LoggerAwareInterface
+{
+    use LoggerAwareTrait;
+
+    /**
+     * Bit mask for blacklist ranges.
+     *
+     * @var int
+     */
+    const IPRANGEFLAG_BLACKLIST = 0x0001;
+
+    /**
+     * Bit mask for whitelist ranges.
+     *
+     * @var int
+     */
+    const IPRANGEFLAG_WHITELIST = 0x0002;
+
+    /**
+     * Bit mask for manually generated ranges.
+     *
+     * @var int
+     */
+    const IPRANGEFLAG_MANUAL = 0x0010;
+
+    /**
+     * Bit mask for automatically generated ranges.
+     *
+     * @var int
+     */
+    const IPRANGEFLAG_AUTOMATIC = 0x0020;
+
+    /**
+     * IP range type: manually added to the blacklist.
+     *
+     * @var int
+     */
+    const IPRANGETYPE_BLACKLIST_MANUAL = 0x0011; // IPRANGEFLAG_BLACKLIST | IPRANGEFLAG_MANUAL
+
+    /**
+     * IP range type: automatically added to the blacklist.
+     *
+     * @var int
+     */
+    const IPRANGETYPE_BLACKLIST_AUTOMATIC = 0x0021; // IPRANGEFLAG_BLACKLIST | IPRANGEFLAG_AUTOMATIC
+
+    /**
+     * IP range type: manually added to the whitelist.
+     *
+     * @var int
+     */
+    const IPRANGETYPE_WHITELIST_MANUAL = 0x0012; // IPRANGEFLAG_WHITELIST | IPRANGEFLAG_MANUAL
+
+    /**
+     * @var \Doctrine\ORM\EntityManagerInterface
+     */
+    protected $em;
+
+    /**
+     * @var \Concrete\Core\Entity\Site\Site
+     */
+    protected $site;
+
+    /**
+     * The IP Access Control Category.
+     *
+     * @var \Concrete\Core\Entity\Permission\IpAccessControlCategory
+     */
+    protected $category;
+
+    /**
+     * @var \IPLib\Address\AddressInterface
+     */
+    protected $defaultIpAddress;
+
+    /**
+     * Initialize the instance.
+     *
+     * @param \Doctrine\ORM\EntityManagerInterface $em
+     * @param \Concrete\Core\Entity\Site\Site $site
+     * @param \Concrete\Core\Entity\Permission\IpAccessControlCategory $category
+     * @param \IPLib\Address\AddressInterface $defaultIpAddress
+     */
+    public function __construct(EntityManagerInterface $em, Site $site, IpAccessControlCategory $category, AddressInterface $defaultIpAddress)
+    {
+        $this->em = $em;
+        $this->site = $site;
+        $this->category = $category;
+        $this->defaultIpAddress = $defaultIpAddress;
+    }
+
+    /**
+     * Get the IP Access Control Category.
+     *
+     * @return \Concrete\Core\Entity\Permission\IpAccessControlCategory
+     */
+    public function getCategory()
+    {
+        return $this->category;
+    }
+
+    /**
+     * Check if an IP address is blacklisted.
+     *
+     * @param \IPLib\Address\AddressInterface|null $ipAddress
+     *
+     * @return bool
+     */
+    public function isBlacklisted(AddressInterface $ipAddress = null)
```

Comment:
@mlocati would it be possible to add `@since` to all public / protected methods? I integrated the IPService in one of my add-ons last week, if I then replace all methods to use this instead, it'll certainly break in anything before 8.5.0. A hint in the code would help a lot against those kind of problems.

Reply:
I don't know which version in @since should be used....